thetis/coupled_timeintegrator.py

```diff
@@ -556,7 +554,9 @@ def advance(self, t, update_forcings=None, update_forcings3d=None):
                 with timed_stage('impl_temp_vdiff'):
                     self.timesteppers.temp_impl.advance(t)
             with timed_stage('impl_mom_vvisc'):
+                self.fields.uv_3d += self.fields.uv_dav_3d
```

Comment:
Yeah, the splitting is quite complicated. For the outputs we hack the 3D field to contain the full velocity, u=\bar{u} + u', in the code it's the deviation u'. (Would probably be better to add another output for the full velocity). The bottom friction term only exists in the 3D momentum equation, so we need to compute that with the full 3D velocity. It's true that for vertical diffusion it does not make any difference if the pronostic variable is u or u'; the vertically constant term has no effect. But it makes a difference for the bottom stress term: The Patankar method guarantees positivity if all sink terms are treated implicitly; you then cannot get negative values. So if you write the vert. mom. equation for u, the implicit stress term guarantees that u does not change sign. If you use u' as the prognostic variable instead, it gets complicated because the stress term is not always a sink (for u').

Reply:
ok, makes sense. Does that mean we should always have use_implicit_vertical_diffusion with this bottom drag formulation? If so, do we need a check for that as it is kind of easy to get wrong?

Reply:
Yes, at the moment you can use this formulation with fully explicit 3D model, but you're right it's hard to imagine a case when that would produce good results.